### DISCUSSION TITLE
Happy to see this project going! Got some feature ideas.

### DISCUSSION THREAD

**Discussion by happyme531** [pine64/OpenPineBuds#23] [Ideas]
I am a headphone user for many years, and there are no open source solution of True Wireless Headset until the PineBuds released.  
That is so amazing. Open source means there will be unique features that is nearly impossible to see in commercial headphones (or only in some overpriced products.)

Some of this features are:

- LDAC Codec.  
  Obviously the only way to implement it in a commercial product is to pay license fee to SONY. 
  But for the open source world [libldacdec](https://github.com/hegdi/libldacdec) is a good choice.

- AptX Codec.
  Only Qualcomm SoCs support it! But wait, really?
  This is more mature, encoding and decoding is supported in GStreamer, Pulseaudio and Pipewire. Plenty of codes and examples.

[cfint](https://github.com/cfint) have added these codecs(also some variants) into ESP32's SDK, ESP-IDF in [its fork](https://github.com/cfint/esp-idf/tree/v4.4-a2dp-sink-codecs-dev). I have tried this and it performs very well. That could be helpful for implementing these into the BES2300 SDK.

- Other Codecs, maybe Opus, FLAC...

- Audio record stream
  Almost all bluetooth headset only support playing music in A2DP profile. For recording, you must switch it to HFP profile, and the audio quality is surprisingly low (comparing to wired microphones).
  [iFLYTEK's iFLYBUDS Pro](https://www.gizmochina.com/2022/05/19/iflyteks-iflybuds-pro-tws-headset-with-50-hours-of-battery-life-launched/)/[Evernote's EverBUDS](https://inf.news/en/news/73e21e86d70bece0538083edfbb4c5b1.html) implemented this feature by (probably) inventing a proprietary protocol to stream the audio to mobile phone with their (also proprietary) apps. 

- Standalone (voice call) audio recording
   Nowadays high-performance bluetooth chips have a large amount of flash memory available. Maybe we can use that to record environment or voice call audio even when the phone is not connected. iFLYBUDS Pro implemented this. Good for Apple users (which does not have a builtin call recording feature).

- Dual connection
  Some really expensive products and some fake Airpods Pros implemented this.
  The headphone can connect two devices simultaneously. But only one device's sound is played at the same time(probably receiving and decoding two streams simultaneously and mix them together requires too much computation power)

- Airpods protocol
  All fake Airpods implemented this. Pretend the device to be a pair of Airpods. Friendly for Apple users.

- 3D rotating sound
  That is probably the easiest to implement: adjusting left and right channel volume by sin(t) and cos(t). No DSP required. Tried on an ESP32, simple but fun.

- More to come...

**Comment by gamelaster** [comment]
Hi! Happy to hear that you like PineBuds Pro :blush: All those ideas are quite realistic to do, although, it will of course take some time, as everyone does this in their free time. But yeah, I agree, there is much potential and I'm excited to see, what community will do!

-gamiee
